polaris-core/src/main/java/org/apache/polaris/core/persistence/cache/EntityCacheGrantManager.java

```diff
@@ -0,0 +1,247 @@
+import org.apache.polaris.core.entity.PolarisGrantRecord;
+import org.apache.polaris.core.entity.PolarisPrivilege;
+import org.apache.polaris.core.persistence.BaseResult;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * PolarisGrantManger implementation that uses an EntityCache to retrieve entities and is backed by
+ * a delegate grant manager for persisting grant changes. This allows consumers to reuse cache
+ * entities without necessarily being aware of the {@link EntityCache} or the {@link
+ * EntityCacheEntry} specifics. Typically, the {@link
+ * org.apache.polaris.core.persistence.resolver.Resolver} is responsible for validating the cache
+ * state. This class does no validation of the entity version or the grant version of any cache
+ * record.
+ */
+public class EntityCacheGrantManager implements PolarisGrantManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(EntityCacheGrantManager.class);
+  private final PolarisGrantManager delegateGrantManager;
+  private final EntityCache entityCache;
+  private PolarisGrantRecord serviceAdminRootContainerGrant;
+  private PolarisBaseEntity serviceAdminEntity;
+
+  public static final class EntityCacheGrantManagerFactory implements PolarisGrantManager.Factory {
+    private final PolarisGrantManager.Factory delegateGrantManagerFactory;
+    private final RealmEntityCacheFactory realmEntityCacheFactory;
+
+    public EntityCacheGrantManagerFactory(
+        Factory delegateGrantManagerFactory, RealmEntityCacheFactory realmEntityCacheFactory) {
+      this.delegateGrantManagerFactory = delegateGrantManagerFactory;
+      this.realmEntityCacheFactory = realmEntityCacheFactory;
+    }
+
+    @Override
+    public PolarisGrantManager getGrantManagerForRealm(RealmContext realm) {
+      return new EntityCacheGrantManager(
+          delegateGrantManagerFactory.getGrantManagerForRealm(realm),
+          realmEntityCacheFactory.getOrCreateEntityCache(realm));
+    }
+  }
+
+  public EntityCacheGrantManager(
+      PolarisGrantManager delegateGrantManager, EntityCache entityCache) {
+    this.delegateGrantManager = delegateGrantManager;
+    this.entityCache = entityCache;
+  }
+
+  @Override
+  public @NotNull PrivilegeResult grantUsageOnRoleToGrantee(
+      @NotNull PolarisCallContext callCtx,
+      @Nullable PolarisEntityCore catalog,
+      @NotNull PolarisEntityCore role,
+      @NotNull PolarisEntityCore grantee) {
+    try {
+      return delegateGrantManager.grantUsageOnRoleToGrantee(callCtx, catalog, role, grantee);
+    } finally {
+      LOGGER.debug("Invalidating cache for role {} and grantee {}", role, grantee);
+      invalidate(role);
+      invalidate(grantee);
+    }
+  }
+
+  private void invalidate(@NotNull PolarisEntityCore role) {
+    EntityCacheEntry roleEntity = entityCache.getEntityById(role.getId());
+    if (roleEntity != null) {
```

Comment:
Why do we need a get before remove (next line)? Why not add remove-by-id method?

Reply:
yeah, that's a good idea. I was just using what was already there :)

Reply:
So ultimately, this does a lookup for the entity in order to generate a cache key anyway. I just moved this logic into the `EntityCache`, so the caller just invokes `removeCacheEntry(PolarisEntityCore)` and it does the lookup and cache key generation internally.